nemo/backends/pytorch/common/multi_data.py

```diff
@@ -0,0 +1,134 @@
+        combination_mode: (DataCombination) defines how to combine the datasets.
+        shuffle: (bool) whether underlying multi dataset should be shuffled in each epoch
+        port_names: List(str) user can override all port names if specified 
+        """
+        super().__init__()
+        self._data_layers = data_layers
+        self._batch_size = batch_size
+        self._shuffle = shuffle
+        self._combination_mode = combination_mode
+        self._port_names = port_names
+        self._dataset = MultiDataset(
+            datasets=[dl.dataset for dl in self._data_layers], combination_mode=combination_mode
+        )
+
+        self._ports = dict()
+        if self._port_names:
+            i = 0
+            for dl in self._data_layers:
+                for _, port_type in dl.output_ports.items():
+                    self._ports[self._port_names[i]] = port_type
+                    i += 1
+        else:
+            for dl_idx, dl in enumerate(self._data_layers):
+                for port_name, port_type in dl.output_ports.items():
+                    if port_name in self._ports:
+                        logging.warning(f"name collision {port_name}, will rename")
+                        self._ports[f"{port_name}_{dl_idx}"] = port_type
+                    else:
+                        self._ports[port_name] = port_type
+
+    @property
+    def output_ports(self):
+        """Return: dict
+        Returns union of all individual data_layer output ports
+        In case of name collision, resolve by renaming 
+        """
+        return self._ports
+
+    def __len__(self):
+        return len(self._dataset)
+
+    @property
+    def dataset(self):
+        return self._dataset
+
+    @property
+    def data_iterator(self):
+        return None
+
+
+class MultiDataset(torch.utils.data.Dataset):
+    def __init__(
+        self,
+        datasets: List[torch.utils.data.Dataset],
+        combination_mode: DataCombination = DataCombination.CROSSPRODUCT,
+    ):
+        """
+        Datasets: list of torch.utils.data.Dataset objects.
+        combination_mode: str, defines how to combine the datasets, Options are ["cross_product", "zip"]. 
+        """
+        self.datasets = datasets
+        self.combination_mode = combination_mode
+        if self.combination_mode == DataCombination.CROSSPRODUCT:
+            self.len = np.prod([len(d) for d in self.datasets])
+        elif self.combination_mode == DataCombination.ZIP:
+            ds_lens = [len(d) for d in self.datasets]
+            self.len = np.min(ds_lens)
+            if len(set(ds_lens)) != 1:
+                raise ValueError("datasets do not have equal lengths.")
+        else:
+            raise ValueError("combination_mode unknown")
+
+    def __getitem__(self, i):
+        """
+        Returns tuple (x1, x2, ...xn) where x1 \in D1, x2 \in D2, ...xn\ Dn
+        """
+
+        return [x for d in self.datasets for x in d[i % len(d)]]
```

Comment:
should you actually return tuple, not list here?

Reply:
should not matter, i prefer list, i can update the documentation to make it consistent

Reply:
yes, just make it consistent